### PR TITLE
Rschmid4/transfer feature

### PR DESCRIPTION
**Pull request by rschmid4**
MVController.java: Modified to display transfer page and backend for sending money.
User.java: Getters and Setters for who you are sending money to, the amount, and a log of transfers
addCustomers.py: Add new transfer log table
welcome.jsp: Modified to have a button to do a transfer
account_info.jsp: modified to display transfer button and transfer logs
transfer_form.jsp: new file for the new page to make a transfer
MVControllerTest.java: Added new tests for the transfer feature.

### CODE REVIEW

**Comment by s1ddh4rthc** [comment]
Opening a new PR with new updates.

**Comment by s1ddh4rthc** [comment]
Added integration tests to test the transfer feature and edge cases. 

Here are the files that were changed along with what changes were made in each of the files:
`/codejava/`: This was an old folder that was changed to testudobank. I removed this directory as it was using an old file organization format that we no longer use.
`MvcController.java`: Commented out the code pertaining to the `TransferHistory` table in the `updateAccountInfo()` because the current `createDB.sql` file did not have a table that was being created. 

Changed some formatting errors that were causing the test to crash earlier such as changing the SQL queries to query the table `Passwords` rather than `passwords`. 

`MvcControllerIntegTest.java`: Created private static variables pertaining to Customer2. This is necessary in order to have a transfer occur, as there is a sender and a recipient required. 

`MvcControllerIntegTest.java --> testTransfer()`: This integration test was written to test the basic functionality of the Transfer Feature. This test initializes two customers with a positive balance and transfers $100 from customer1's account to customer2. This test makes sure that the new balances are accurately changed by making sure customer1's balance decreases by the transfer amount and customer2's balance increases by the specified amount.

`MvcControllerIntegTest.java --> testTransferPaysOffOverdraftBalance()`: This integration test tests the scenario where the recipient of a transfer has a current overdraft balance, and the transfer amount pays off some of the overdraft. In this test, the recipient, `customer2`, has an overdraft balance of $101, and the transfer amount is $100. This test makes sure that the user has a remaining overdraft balance of $1. 

'MvcControllerIntegTest.java --> testTransferPaysOverdraftAndDepositsRemainder()`: This integration test was written for the scenario in which the recipient has an overdraft balance and the transfer amount is greater than their current overdraft. This test makes sure that the remainder of the balance post overdraft payments are added to the recipient's balance. 

'createDB.sql': Changed this file to create a new table called TransferHistory for users in MySQL.

**Comment by s1ddh4rthc** [comment]
`src/main/java/net/testudobank/MvcController.java`:
Changed previous code to reflect the current code paradigms and format that we're using. Added some comments in order to provide greater clarity as to what is happening in the code. 

`src/main/java/net/testudobank/TestudoBankRepository.java`:
Added two methods in order to simplify the code within `submitTransfer()` and in case future functionality requires methods like these.
`doesCustomerExist()`: Checks whether a customer exists by checking if their CustomerID is found in the `Customers` table. If not, returns "welcome", our default error handler.

`insertRowToTransferLogsTable`: Due to a new TransferHistory table being created, created this method to help simplify the process of adding entries to the TransferHistory table.

File: src/main/java/net/testudobank/MvcController.java [hunk 1]
```diff
@@ -328,6 +342,140 @@ public String submitWithdraw(@ModelAttribute("user") User user) {
...
+    }
+
+    //checks to see the customer you are transfering to exists
+    TestudoBankRepository.doesCustomerExist(jdbcTemplate, user.getWhoToTransfer());
+    
+    double userWithdrawAmt = user.getAmountToTransfer();
+    int userWithdrawAmtInPennies = convertDollarsToPennies(userWithdrawAmt);
+
+    //Get the sender's balance in pennies.
+    int userBalanceInPennies = TestudoBankRepository.getCustomerBalanceInPennies(jdbcTemplate, userID);
```
Comment (adithyasolai) [inline src/main/java/net/testudobank/MvcController.java:406]: Unnecessary comment. Just rename the variable to `senderBalanceInPennies`, and remove the comment. The code should speak for itself.

File: src/main/java/net/testudobank/TestudoBankRepository.java [hunk 2]
```diff
@@ -102,6 +112,14 @@ public static void deleteRowFromOverdraftLogsTable(JdbcTemplate jdbcTemplate, St
     jdbcTemplate.update(deleteRowFromOverdraftLogsSql);
   }
 
+  public static void insertRowToTransferLogsTable(JdbcTemplate jdbcTemplate, String customerID, String recipientID, int transferAmount) {
```
Comment (adithyasolai) [inline src/main/java/net/testudobank/TestudoBankRepository.java:115]: This is a useful helper method that can be kept for the future.

**Review by adithyasolai — CHANGES_REQUESTED** [review]
Lots of things to fix. Lots of important discussions we need to have from a design standpoint on how to move forward. I can foresee a new DB Schema change. Look at comments below for details.

File: src/main/java/net/testudobank/MvcController.java [hunk 3]
```diff
@@ -131,6 +131,19 @@ private void updateAccountInfo(User user) {
...
+    // List<Map<String,Object>> transferLogsFrom = jdbcTemplate.queryForList(getTransferLogsFromSql);
+    // String transferLogsToOutput = HTML_LINE_BREAK;
+    // for(Map<String, Object> transferLogTo : transferLogsTo){
+    //   transferLogsToOutput += transferLogTo + HTML_LINE_BREAK;
+    // }
+    // String transferLogsFromOutput = HTML_LINE_BREAK;
+    // for(Map<String, Object> transferLogFrom : transferLogsFrom){
+    //   transferLogsFromOutput += transferLogFrom + HTML_LINE_BREAK;
+    // }
+
```
[the remark this answers is not in the document]
Comment (s1ddh4rthc) [reply src/main/java/net/testudobank/MvcController.java:146]: My bad, didn't see that I left this commented. Uncommented this block.

File: src/main/java/net/testudobank/MvcController.java [hunk 4]
```diff
@@ -142,6 +155,7 @@ private void updateAccountInfo(User user) {
     user.setOverDraftBalance(overDraftBalance/100);
     user.setLogs(logs);
     user.setTransactionHist(transactionHistoryOutput);
+    //user.setTransferLogs(transferLogsToOutput + transferLogsFromOutput);
```
[the remark this answers is not in the document]
Comment (s1ddh4rthc) [reply src/main/java/net/testudobank/MvcController.java:158]: Uncommented.

File: src/main/java/net/testudobank/MvcController.java [hunk 5]
```diff
@@ -328,6 +342,140 @@ public String submitWithdraw(@ModelAttribute("user") User user) {
 
   }
 
+/**
+   * HTML GET request handler that serves the "transfer_form" page to the user.
+   * An empty `User` object is also added to the Model as an Attribute to store
+   * the user's transfer form input.
+   * 
+   * @param model
+   * @return "dispute_form" page
```
[the remark this answers is not in the document]
Comment (s1ddh4rthc) [reply src/main/java/net/testudobank/MvcController.java:351]: Done.

File: src/main/java/net/testudobank/MvcController.java [hunk 6]
```diff
@@ -328,6 +342,140 @@ public String submitWithdraw(@ModelAttribute("user") User user) {
...
+   * 
+   * The same username+password handling from the login page is used.
+   * 
+   * If the password attempt is correct, the users transfer successfully goes through
+   * if it is a valid transfer. Both customers balances are properly updated.
+   * 
+   * If the password attempt is incorrect, the user is redirected to the "welcome" page.
+   * 
+   * @param user
+   * @return "account_info" page if login successful. Otherwise, redirect to "welcome" page.
```
[the remark this answers is not in the document]
Comment (s1ddh4rthc) [reply src/main/java/net/testudobank/MvcController.java:371]: Done.

File: src/main/java/net/testudobank/MvcController.java [hunk 7]
```diff
@@ -328,6 +342,140 @@ public String submitWithdraw(@ModelAttribute("user") User user) {
...
+    int numOfReversals = TestudoBankRepository.getCustomerNumberOfReversals(jdbcTemplate, userID);
+    if (numOfReversals >= MAX_DISPUTES || user.getWhoToTransfer().equals(userID)) {
+      return "welcome";
+    }
+
+    // Negative transfer amount is not allowed
+    double userDepositAmt = user.getAmountToTransfer();
+    if (userDepositAmt < 0) {
+      return "welcome";
+    }
```
[the remark this answers is not in the document]
Comment (s1ddh4rthc) [reply src/main/java/net/testudobank/MvcController.java:397]: Sure, I've left them for now, but we can revisit this. I agree with the fact that it is more readable with the current structure.

File: src/main/java/net/testudobank/MvcController.java [hunk 8]
```diff
@@ -328,6 +342,140 @@ public String submitWithdraw(@ModelAttribute("user") User user) {
...
+    TestudoBankRepository.doesCustomerExist(jdbcTemplate, user.getWhoToTransfer());
+    
+    double userWithdrawAmt = user.getAmountToTransfer();
+    int userWithdrawAmtInPennies = convertDollarsToPennies(userWithdrawAmt);
+
+    //Get the sender's balance in pennies.
+    int userBalanceInPennies = TestudoBankRepository.getCustomerBalanceInPennies(jdbcTemplate, userID);
+    
+    // if the balance is not positive, withdraw with interest fee
+    if (userWithdrawAmtInPennies > userBalanceInPennies) {
```
[the remark this answers is not in the document]
Comment (s1ddh4rthc) [reply src/main/java/net/testudobank/MvcController.java:409]: Good note. I've changed the variables to reflect these changes.

File: src/main/java/net/testudobank/TestudoBankRepository.java [hunk 9]
```diff
@@ -48,6 +48,16 @@ public static int getCustomerOverdraftBalanceInPennies(JdbcTemplate jdbcTemplate
...
+  public static String doesCustomerExist(JdbcTemplate jdbcTemplate, String customerID) { 
+    String getCustomerIDSql =  String.format("SELECT CustomerID FROM Customers WHERE CustomerID='%s';", customerID);
+    try {
+      return jdbcTemplate.queryForObject(getCustomerIDSql, String.class);
+    }
+    catch(Exception e) {
+      return "welcome";
+    }
+  }
+
```
[the remark this answers is not in the document]
Comment (s1ddh4rthc) [reply src/main/java/net/testudobank/TestudoBankRepository.java:60]: Removed it and put this code back only in the `submitTransfer()` method.

File: src/main/java/net/testudobank/TestudoBankRepository.java (hunk 2 above)
Comment (s1ddh4rthc) [reply src/main/java/net/testudobank/TestudoBankRepository.java:115]: Cool. Won't remove this then.

File: src/test/java/net/testudobank/tests/MvcControllerTest.java [hunk 10]
```diff
@@ -333,5 +335,234 @@ public void testWithdrawDisputeInOverdraft() {
     Mockito.verify(jdbcTemplate, Mockito.times(1)).update(eq(overdraftBalanceUpdateSql));
     // verify "account_info" page is returned
 		assertEquals("account_info", pageReturned);
+  }
+  
+  //@Test
+	public void testTransfer() {
```
[the remark this answers is not in the document]
Comment (s1ddh4rthc) [reply src/test/java/net/testudobank/tests/MvcControllerTest.java:341]: Yes, I did not write any code in the `MvcControllerTest.java` file.

File: src/main/java/net/testudobank/MvcController.java [hunk 11]
```diff
@@ -328,6 +342,140 @@ public String submitWithdraw(@ModelAttribute("user") User user) {
...
+    }
+
+    //If a customer already has too many reversals, or they try to send money to themselves, transfer fails.
+    int numOfReversals = TestudoBankRepository.getCustomerNumberOfReversals(jdbcTemplate, userID);
+    if (numOfReversals >= MAX_DISPUTES || user.getWhoToTransfer().equals(userID)) {
+      return "welcome";
+    }
+
+    // Negative transfer amount is not allowed
+    double userDepositAmt = user.getAmountToTransfer();
```
[the remark this answers is not in the document]
Comment (s1ddh4rthc) [reply src/main/java/net/testudobank/MvcController.java:394]: Changed.

File: src/test/java/net/testudobank/tests/MvcControllerTest.java (hunk 10 above)
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerTest.java:341]: Delete this unit test because you just implemented an integ test for it.

File: src/test/java/net/testudobank/tests/MvcControllerTest.java [hunk 12]
```diff
@@ -333,5 +335,234 @@ public void testWithdrawDisputeInOverdraft() {
...
+    String balanceIncreaseSqlCustomer2 = String.format("UPDATE Customers SET Balance = Balance + %d WHERE CustomerID='%s';",
+                                                        customer1.getAmountToTransfer() * 100,
+                                                        customer1.getWhoToTransfer());
+    Mockito.verify(jdbcTemplate, Mockito.times(1)).update(eq(balanceIncreaseSqlCustomer2));
+    // verify "account_info" page is returned
+		assertEquals("account_info", pageReturned);
+	}
+
+  //@Test
+	public void testTransferPayOffOverdraft() {
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerTest.java:399]: Delete this unit test because you just implemented an integ test for it.

File: src/test/java/net/testudobank/tests/MvcControllerTest.java [hunk 13]
```diff
@@ -333,5 +335,234 @@ public void testWithdrawDisputeInOverdraft() {
...
+    String balanceIncreaseSqlCustomer2 = String.format("UPDATE Customers SET Balance = Balance + %d WHERE CustomerID='%s';",
+                                                        90,
+                                                        customer1.getWhoToTransfer());
+    Mockito.verify(jdbcTemplate, Mockito.times(1)).update(eq(balanceIncreaseSqlCustomer2));
+    // verify "account_info" page is returned
+		assertEquals("account_info", pageReturned);
+	}
+
+  @Test
+	public void testTransferExceedsOverdraftLimit() {
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerTest.java:460]: Write an integ test for this case.

File: src/test/java/net/testudobank/tests/MvcControllerTest.java [hunk 14]
```diff
@@ -333,5 +335,234 @@ public void testWithdrawDisputeInOverdraft() {
...
+    String balanceIncreaseSqlCustomer2 = String.format("UPDATE Customers SET Balance = Balance + %d WHERE CustomerID='%s';",
+                                                        customer1.getAmountToTransfer()*100,
+                                                        customer1.getWhoToTransfer());
+    Mockito.verify(jdbcTemplate, Mockito.times(0)).update(eq(balanceIncreaseSqlCustomer2));
+    // verify "welcome" page is returned
+		assertEquals("welcome", pageReturned);
+	}
+
+  @Test
+	public void testTransferToYourself() {
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerTest.java:518]: Write an integ test for this error case.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 15]
```diff
@@ -591,6 +596,201 @@ public void testReversalOfSimpleWithdraw() throws SQLException, ScriptException,
...
+    int CUSTOMER1_BALANCE_IN_PENNIES = MvcControllerIntegTestHelpers.convertDollarsToPennies(CUSTOMER1_BALANCE);
+    MvcControllerIntegTestHelpers.addCustomerToDB(dbDelegate, CUSTOMER1_ID, CUSTOMER1_PASSWORD, CUSTOMER1_FIRST_NAME, CUSTOMER1_LAST_NAME, CUSTOMER1_BALANCE_IN_PENNIES);
+
+    //Initialize customer2 with a balance of $500. Balance will be represented as pennies in DB. 
+    double CUSTOMER2_BALANCE = 500;
+    int CUSTOMER2_BALANCE_IN_PENNIES = MvcControllerIntegTestHelpers.convertDollarsToPennies(CUSTOMER2_BALANCE);
+    MvcControllerIntegTestHelpers.addCustomerToDB(dbDelegate, CUSTOMER2_ID, CUSTOMER2_PASSWORD, CUSTOMER2_FIRST_NAME, CUSTOMER2_LAST_NAME, CUSTOMER2_BALANCE_IN_PENNIES);
+
+    //Amount to transfer
+    int TRANSFER_AMOUNT = 100;
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:622]: Make this a `double` because the front-end input takes all dollar amounts which can include a decimal. Even though this is a simple whole-dollar amount, it logically should still be a `double` to help the reader understand the logic of why we need to convert to pennies.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 16]
```diff
@@ -591,6 +596,201 @@ public void testReversalOfSimpleWithdraw() throws SQLException, ScriptException,
...
+    User CUSTOMER1 = new User();
+    CUSTOMER1.setUsername(CUSTOMER1_ID);
+    CUSTOMER1.setPassword(CUSTOMER1_PASSWORD);
+    CUSTOMER1.setWhoToTransfer(CUSTOMER2_ID);
+    CUSTOMER1.setAmountToTransfer(TRANSFER_AMOUNT);
+
+    //Send the transfer request.
+    String returnedPage = controller.submitTransfer(CUSTOMER1);
+    
+    //Fetch customer1's data from DB
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:635]: make this `// fetch customer1 & customer2's data from DB`

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 17]
```diff
@@ -591,6 +596,201 @@ public void testReversalOfSimpleWithdraw() throws SQLException, ScriptException,
...
+   int CUSTOMER1_BALANCE_IN_PENNIES = MvcControllerIntegTestHelpers.convertDollarsToPennies(CUSTOMER1_BALANCE);
+   MvcControllerIntegTestHelpers.addCustomerToDB(dbDelegate, CUSTOMER1_ID, CUSTOMER1_PASSWORD, CUSTOMER1_FIRST_NAME, CUSTOMER1_LAST_NAME, CUSTOMER1_BALANCE_IN_PENNIES);
+
+   //Initialize customer2 with a balance of $100. Balance will be represented as pennies in DB. 
+   double CUSTOMER2_BALANCE = 100;
+   int CUSTOMER2_BALANCE_IN_PENNIES = MvcControllerIntegTestHelpers.convertDollarsToPennies(CUSTOMER2_BALANCE);
+   MvcControllerIntegTestHelpers.addCustomerToDB(dbDelegate, CUSTOMER2_ID, CUSTOMER2_PASSWORD, CUSTOMER2_FIRST_NAME, CUSTOMER2_LAST_NAME, CUSTOMER2_BALANCE_IN_PENNIES);
+
+   //Amount to transfer
+   int TRANSFER_AMOUNT = 100;
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:676]: Make this a `double`.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 18]
```diff
@@ -591,6 +596,201 @@ public void testReversalOfSimpleWithdraw() throws SQLException, ScriptException,
...
+    List<Map<String,Object>> customer1SqlResult = jdbcTemplate.queryForList(String.format("SELECT * FROM Customers WHERE CustomerID='%s';", CUSTOMER1_ID));
+    Map<String, Object> customer1Data = customer1SqlResult.get(0);
+
+    List<Map<String,Object>> customer2SqlResult = jdbcTemplate.queryForList(String.format("SELECT * FROM Customers WHERE CustomerID='%s';", CUSTOMER2_ID));
+    Map<String, Object> customer2Data = customer2SqlResult.get(0);
+
+    //Verify that customer1's balance decreased by $100. 
+    assertEquals((CUSTOMER1_BALANCE_IN_PENNIES - TRANSFER_AMOUNT_IN_PENNIES), (int)customer1Data.get("Balance"));
+
+    //Verify that customer2's overdraft balance decreased by $100.
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:712]: Also verify that customer2's main balance is still $0, which is what makes this different from your next integ test below.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 19]
```diff
@@ -591,6 +596,201 @@ public void testReversalOfSimpleWithdraw() throws SQLException, ScriptException,
...
+   int CUSTOMER2_EXPECTED_OVERDRAFT_BALANCE = MvcControllerIntegTestHelpers.applyOverdraftInterest(CUSTOMER2_BALANCE_IN_PENNIES - CUSTOMER2_AMOUNT_TO_WITHDRAW_IN_PENNIES) * -1;
+   CUSTOMER2.setUsername(CUSTOMER2_ID);
+   CUSTOMER2.setPassword(CUSTOMER2_PASSWORD);
+   CUSTOMER2.setAmountToWithdraw(CUSTOMER2_AMOUNT_TO_WITHDRAW);
+
+   CUSTOMER1.setWhoToTransfer(CUSTOMER2_ID);
+   CUSTOMER1.setAmountToTransfer(TRANSFER_AMOUNT);
+
+    //Withdraw $201 from Customer2's account to cause an overdraft of $101.
+    controller.submitWithdraw(CUSTOMER2);
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:697]: No need to do these withdraws to cause customer2 to go into overdraft. You can make customer2 start in the overdraft state.

You can just use the other `addCustomerToDB()` helper method that takes more inputs to initialize customer2 with a main balance of $0 and overdraft balance of $101. Look at the `testDepositOverdraftClearedWithExcess()` integ test for an example.

File: src/test/java/net/testudobank/tests/MvcControllerIntegTest.java [hunk 20]
```diff
@@ -591,6 +596,201 @@ public void testReversalOfSimpleWithdraw() throws SQLException, ScriptException,
...
+   int CUSTOMER2_EXPECTED_OVERDRAFT_BALANCE = MvcControllerIntegTestHelpers.applyOverdraftInterest(CUSTOMER2_BALANCE_IN_PENNIES - CUSTOMER2_AMOUNT_TO_WITHDRAW_IN_PENNIES) * -1;
+   CUSTOMER2.setUsername(CUSTOMER2_ID);
+   CUSTOMER2.setPassword(CUSTOMER2_PASSWORD);
+   CUSTOMER2.setAmountToWithdraw(CUSTOMER2_AMOUNT_TO_WITHDRAW);
+
+   CUSTOMER1.setWhoToTransfer(CUSTOMER2_ID);
+   CUSTOMER1.setAmountToTransfer(TRANSFER_AMOUNT);
+
+   //Withdraw $200 from Customer2's account to cause an overdraft of $100.
+   controller.submitWithdraw(CUSTOMER2);
```
Comment (adithyasolai) [inline src/test/java/net/testudobank/tests/MvcControllerIntegTest.java:767]: Again, you can just initialize customer2 to have an overdraft balance of $102 without doing these withdraws.

File: src/main/java/net/testudobank/User.java [hunk 21]
```diff
@@ -1,48 +1,57 @@
...
+	private String transferLogs;
+
+  @Setter @Getter
+  private int numTransactionsAgo;
+
+  @Setter @Getter
+  private int amountToTransfer;
+
+  @Setter @Getter
+	private String whoToTransfer;
```
Comment (adithyasolai) [inline src/main/java/net/testudobank/User.java:50]: Change this to a better variable name like `transferRecipientID`

File: src/main/java/net/testudobank/MvcController.java (hunk 1 above)
Comment (adithyasolai) [reply src/main/java/net/testudobank/MvcController.java:411]: Still need to implement this.

File: src/test/resources/createDB.sql [hunk 22]
```diff
@@ -26,3 +26,9 @@ CREATE TABLE TransactionHistory (
   Action varchar(255) CHECK (Action IN ('Deposit', 'Withdraw')),
   Amount int
 );
+
+CREATE TABLE TransferHistory (
+  TransferFrom varchar(255),
+  TransferTo varchar(255),
+  Amount int
+);
```
Comment (adithyasolai) [inline src/test/resources/createDB.sql:34]: @jacktzhao and I will need to add a `Timestamp` to this table later as per the new re-design.

File: src/test/resources/createDB.sql [hunk 23]
```diff
@@ -26,3 +26,9 @@ CREATE TABLE TransactionHistory (
   Action varchar(255) CHECK (Action IN ('Deposit', 'Withdraw')),
```
Comment (adithyasolai) [inline src/test/resources/createDB.sql:26]: @jacktzhao and I will need to add `Transfer` as a valid action here as per the new re-design.

**Review by adithyasolai — CHANGES_REQUESTED** [review]
Added a few more things you can fix up without working on the whole re-design. Adding new integ tests shouldn't hurt us even if we change the Transfer logic in a bit to match the re-design. We will likely need to add a few more checks to the existing Transfer integ tests after doing the re-design.
